React header

fhlavac opened the pull request:
@Hyperkid123 @martinpovolny @epwinchell

- Added parameter for element specification to ReactjsHelper
- Rewritten help and user menu structure generating (similar as for vertical navbar - should be merged)
- Rewritten top-navbar component and divided it into components: user-options, configuration, help, notifications, custom-logo
- Created right-section component containing all above mentioned small components
- Created top-navbar tests

This PR is dependent on https://github.com/ManageIQ/manageiq-ui-classic/pull/5854

![top-navbar](https://user-images.githubusercontent.com/50696716/65047113-29a1c200-d962-11e9-810d-e792bf714ac7.png)

In notifications part I currently call setUnreadCount (function returning number of unread notifications) by Redux. Previously, it was solved in Angular. 
Currently: When number of unread notifications in drawer is updated, Rx message is sent to React header and Redux dispatch updates value stored in Redux store.

epwinchell commented:
@miq-bot add_reviewer @martinpovolny

epwinchell commented:
@miq-bot add_reviewer @epwinchell

epwinchell commented:
@miq-bot add_reviewer @Hyperkid123

martinpovolny commented:
Touches notifications, pinging @skateman .

Also the notifications should be converted to React as a priority.

This is a piece of Angular that is loaded on every page. Let's avoid that soon. Toolbars will be gone in a moment, then GTL will be replaced and trees. There should be no angular on most pages in the next release.

skateman commented:
@martinpovolny the pf-react implementation of the drawer is very far from our needs. We did some customizations with @himdel on top of the angular-based implementation and we would have to do that and a little more with the react one before we're able to actually consume it.

I was looking into this last year and it seemed like a painful task. But I'm in the area already because of the popup throttling, so I'll re-evaluate this as well. 

Gonna take a look at this piece of code of course ... gimme some :clock1:

skateman commented on app/assets/stylesheets/main.scss:
```diff
@@ -66,5 +66,4 @@ $login-container-details-border-color-rgba: rgba(0, 0, 0, 0.5);
   background-size: 100% auto;
-}
-
+}
```
Missing newline here...

skateman commented on app/helpers/application_helper/navbar.rb:
```diff
@@ -1,5 +1,33 @@
+        :items          => item.items.to_a.map(&method(:item_to_hash)),
+        :visible        => item.visible?,
+        :link_params    => item.link_params
+      }
```
You could implement this a little prettier:
```ruby
%i[
  id
  name
  icon
  placement
  before
  type
  href
  parent_id
  feature
  rbac_feature
  defaults
  link_params
].each_with_object(:visible => item.visible?, :items => item.items.to_a.map(&method(:item_to_hash))) do |obj, key|
  obj[key] = item.try(key)
end
```

Except of `visible` and `items` all the fields map nicely to the keys...

skateman replied:
10 points to Griffindor for the `&method` btw :1st_place_medal: :laughing:

skateman commented on app/javascript/components/top-navbar/right-section.jsx:
```diff
@@ -0,0 +1,63 @@
+    <Help helpMenu={helpMenu} />
+    { currentUser && (
+      <React.Fragment>
+        <Configuration opsExplorerAllowed={opsExplorerAllowed} />
```
TBH I don't see a reason why to have this as a separate component in a separate file, this is way too small for that and you're creating much bigger ones here. I'd just extract it into a function, but don't take this seriously and ask someone who actually knows React :wink:

Hyperkid123 replied:
This was actually by my command! :skull_and_crossbones: 

But it is better to make these component as small and independent as possible. It is much easier to find them, test them and fix them if need be.

skateman replied:
Then IMO the other one(s) should be small(er) as well... It's just feels weird that you have one small, and the rest is HUGE.

skateman commented on app/javascript/components/top-navbar/right-section.jsx:
```diff
@@ -0,0 +1,63 @@
+  customLogo, currentUser, helpMenu, opsExplorerAllowed, applianceName, miqGroups, currentGroup, userMenu,
+}) => (
+  <React.Fragment>
+    <CustomLogo customLogo={customLogo} />
```
Same as below...

skateman commented on app/assets/javascripts/controllers/header/header_controller.js:
```diff
@@ -47,6 +53,7 @@ angular.module('ManageIQ').controller('headerController', ['$scope', 'eventNotif
 
     vm.notificationsIndicatorTooltip = miqFormatNotification(__('%{count} unread notifications'),
       {count: notificationCount});
+      ManageIQ.redux.store.dispatch({ type: '@@notifications/setUnreadCount', payload: notificationCount.text });
```
Could you please write a few words into the PR's description about this idea. I know how it works because I made it up :laughing: but others might not...

skateman commented on app/assets/javascripts/services/event_notifications_service.js:
```diff
@@ -91,6 +91,11 @@ angular.module('ManageIQ').service('eventNotifications', ['$timeout', '$window',
 
           updateUnreadCount(events);
+          var x = 0;
+          _.forEach(state.groups, function(group) {
```
I think you don't need the `_` here, you [can](https://caniuse.com/#feat=mdn-javascript_builtins_array_foreach) directly call [`forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) on the array. Anyway, this code is a great example for using [`reduce`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce) on the array instead of having the helper variable.

skateman commented on app/helpers/application_helper/navbar.rb:
```diff
@@ -1,5 +1,33 @@
 module ApplicationHelper
   module Navbar
+    def menu_to_json(position)
```
For now, it's okay to have this method here, but after this is done, we should extract it into the `Menu::Manager`. I have an idea about that, just let's not forget it so we need to talk when this is done and we're both in the office.

skateman commented on app/views/layouts/_right_section.html.haml:
```diff
@@ -0,0 +1 @@
+= react('RightSection', { :customLogo => ::Settings.server.custom_logo, :currentUser => current_user, :helpMenu => menu_to_json(:help),:opsExplorerAllowed => ApplicationHelper.role_allows?(:feature => 'ops_explorer', :any => true), :applianceName => appliance_name, :miqGroups => current_user.miq_groups, :currentGroup => current_group, :userMenu => menu_to_json(:top_right) }, {:class => "nav navba ...
```
Missing newline

skateman replied:
Maybe it's completely wrong and maybe out of the context of this PR, but could we preseed these data into the Redux store upon login? Then we would have them cached and the list of props would be simpler/shorter/nicer...

skateman commented on app/javascript/miq-redux/store.js:
```diff
@@ -7,6 +7,21 @@ const initialState = {};
+  unreadCount: 0,
+};
+
+const nPrefix = '@@notifications/';
```
It's just a semantic nitpick and others might disagree with me, but I'd prefer singular here. Just to be a little more future-prooof. Also the `setUnreadCount` is a little too specific and there's a 100% chance it will be dropped in the future in favor of something like `new`. So I'd just go with `@@notification/new` instead, but this is also something where others might and should stop me.

Hyperkid123 replied:
Honestly i don't really care about this :D As long as its same everywhere I am ok with both.

skateman replied:
Yeah, in this case let's go with singular please.

Hyperkid123 commented on app/assets/javascripts/services/event_notifications_service.js:
```diff
@@ -91,6 +91,8 @@ angular.module('ManageIQ').service('eventNotifications', ['$timeout', '$window',
           });
 
           updateUnreadCount(events);
+          const totalUnreadCount = state.groups.reduce((sum, group) => (sum + group.unreadCount), 0);
```
I don't think that we can use `Array.prototype.reduce` in old JS. This is not ES6 and its not being compiled by webpack.

himdel replied:
es6-shim. We can use any ES6 libary feature just fine, only the syntax ones have to go through babel.

himdel replied:
@fhlavac this is the one ^.. `reduce` is fine, but arrow functions are new *syntax*

himdel replied:
Oh, and also `const`! :)

Hyperkid123 commented on app/javascript/components/top-navbar/configuration.jsx:
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+
+const Configuration = ({
```
I don't see any prop types validation here.

Hyperkid123 commented on app/javascript/components/top-navbar/help.jsx:
```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import { Dropdown, Icon, MenuItem } from 'patternfly-react';
+
+const Help = ({
```
Missing prop types again.

Hyperkid123 commented on app/javascript/components/top-navbar/user-options.jsx:
```diff
@@ -0,0 +1,110 @@
+import React from 'react';
+import { Dropdown, Icon, MenuItem } from 'patternfly-react';
+
+const UserOptions = ({
```
PropTypes

Hyperkid123 commented on app/javascript/packs/component-definitions-common.js:
```diff
@@ -70,6 +76,12 @@ ManageIQ.component.addReact('TagGroup', props => <TagGroup {...props} />);
 ManageIQ.component.addReact('TagView', TagView);
 ManageIQ.component.addReact('TaggingWrapperConnected', TaggingWrapperConnected);
 ManageIQ.component.addReact('TextualSummaryWrapper', TextualSummaryWrapper);
+ManageIQ.component.addReact('RightSection', RightSection);
```
These should be sorted alphabetically by the component name. It helps us to avoid constant PR conflicts when registering new react components.

martinpovolny commented:
Looks good!

Please, do not forget to move `menu_to_json`.
I think it's ok to have small and big components either in one or in multiple files. It really does not bother me for small React components.

Thx, @fhlavac, @Hyperkid123, @skateman !

himdel commented on app/javascript/components/top-navbar/user-options.jsx:
```diff
@@ -0,0 +1,139 @@
+                  id={group.description}
+                  key={group.id}
+                  title={__('Change to this Group')}
+                  onClick={(e) => { e.preventDefault(); miqSparkleOn(); miqToggleUserOptions(group.id); }}
```
This will need a change once https://github.com/ManageIQ/manageiq-ui-classic/pull/6325 gets merged.

`e.preventDefault(); miqChangeGroup(group.id);` should do it, just make sure the id is a string please ;)

himdel commented on app/javascript/components/top-navbar/user-options.jsx:
```diff
@@ -0,0 +1,139 @@
+      <MenuItem
+        id="logout-btn"
+        href="/dashboard/logout"
+        onClick={event => !miqCheckForChanges() && event.preventDefault()}
```
And this also needs updating - this time for https://github.com/ManageIQ/manageiq-ui-classic/pull/6065

I suggest

```
(e) => miqCheckForChanges() ? (ManageIQ.logoutInProgress = true) : e.preventDefault()
```
```

Hyperkid123 commented:
@himdel there seems to be something very wrong travis? Do you have an idea if so much failures could be caused by replacing the header menu? I don't think so (i also can't restart the build because travis kicked me out of all organizations).

@fhlavac did rebased this branch on master? IF not please do that and push it again.

fhlavac commented:
@Hyperkid123 yesterday I rebased it on master and pushed. I can try it again next week at work.

skateman commented:
Probably not caused by you, but for the record:
![Screenshot from 2019-11-15 10-09-06](https://user-images.githubusercontent.com/649130/68931049-017ee580-0790-11ea-87ea-594390250e78.png)


BUT I also found an issue, even if the help menu's `ManageIQ.org` is set by default to open in a new window (target=_blank), it doesn't ... but I'm not able to reproduce in master.
![Screenshot from 2019-11-15 10-14-02](https://user-images.githubusercontent.com/649130/68931416-c4ffb980-0790-11ea-81c6-1208f3ce4c7b.png)

himdel commented:
@Hyperkid123 Oh this can absolutely cause this many failures :D 

The key one is hidden in spec:javascript:
#<ExecJS::RuntimeError: SyntaxError: Unexpected token: operator (>) (line: 18111, col: 69, pos: 6341263)

Using new JS in old places has weird effects, I *think* the ruby failures might disappear after that :)

Hyperkid123 commented:
@himdel lol. I missed that again.

skateman commented on app/javascript/components/top-navbar/help.jsx:
```diff
@@ -50,6 +50,7 @@ const Help = ({
                 key={item.id}
                 href={item.type === 'modal' ? '' : item.href}
                 onClick={e => (item.type === 'modal' ? showAboutModal(e) : !miqCheckForChanges() && e.preventDefault())}
+                target={item.type === 'new_window' ? '_blank' : '_current'}
```
Try using  `getTargetByType` here, to be consistent with the other places.

Hyperkid123 commented on app/javascript/components/top-navbar/notifications.jsx:
```diff
@@ -0,0 +1,31 @@
+      }}
+    >
+      <Icon type="fa" name="bell" />
+      <span className="badge badge-pf-bordered">{unreadCount > 0 ? ' ' : ''}</span>
```
Just want to add note here for anyone that might think this is strange. This is actually how you have to trigger the blue circle above the icon. It checks "empty" css selector to make the decision to hide/show the circle.

Hyperkid123 commented on app/javascript/components/top-navbar/right-section.jsx:
```diff
@@ -0,0 +1,58 @@
+);
+
+RightSection.defaultProps = {
+  currentUser: null,
```
Is this even possible? Can the user prop be empty?

Hyperkid123 commented on app/javascript/components/top-navbar/user-options.jsx:
```diff
@@ -0,0 +1,148 @@
+/* global miqToggleUserOptions miqCheckForChanges */
```
I think these setting should be in the `/app/javascript/.eslintrc` to set these rules globally. I can imagine that more files would have similar issues.

Hyperkid123 commented on app/javascript/components/top-navbar/user-options.jsx:
```diff
@@ -0,0 +1,148 @@
+);
+
+UserOptions.defaultProps = {
+  currentUser: null,
```
Again i think the current user should be required prop.

himdel commented on app/javascript/components/top-navbar/recursive-props.js:
```diff
@@ -0,0 +1,43 @@
+
+export const groupProps = {
+  description: PropTypes.string.isRequired,
+  id: PropTypes.number.isRequired,
```
`id` must be a string

himdel commented on app/javascript/components/top-navbar/user-options.jsx:
```diff
@@ -0,0 +1,142 @@
+                  title={__('Change to this Group')}
+                  onClick={(e) => {
+                    e.preventDefault();
+                    miqChangeGroup(group.id.toString());
```
This bit won't work, `group.id` must already be a string => https://github.com/ManageIQ/manageiq-ui-classic/pull/6197/files#r349078197

himdel commented on app/views/layouts/_right_section.html.haml:
```diff
@@ -0,0 +1 @@
+= react('RightSection', { :customLogo => ::Settings.server.custom_logo, :currentUser => current_user, :helpMenu => menu_to_json(:help),:opsExplorerAllowed => ApplicationHelper.role_allows?(:feature => 'ops_explorer', :any => true), :applianceName => appliance_name, :miqGroups => current_user.miq_groups, :currentGroup => current_group, :userMenu => menu_to_json(:top_right) }, {:class => "nav navba ...
```
split this into 1 parameter per line please :)

And you'll need a map over `miqGroups` to `to_s` the ids

himdel commented on app/javascript/spec/top-navbar/top-navbar.spec.js:
```diff
@@ -0,0 +1,214 @@
+    applianceName = 'EVM';
+    miqGroups = [
+      {
+        id: 10000000000002,
```
(also strings)

himdel commented on app/javascript/packs/component-definitions-common.js:
```diff
@@ -37,6 +37,12 @@ import SetOwnershipForm from '../components/set-ownership-form';
 import TableListViewWrapper from '../react/table_list_view_wrapper';
 import TaggingWrapperConnected from '../components/taggingWrapper';
 import TextualSummaryWrapper from '../react/textual_summary_wrapper';
+import RightSection from '../components/top-navbar/right-section';
```
Should we prefix these components names somehow?

I don't think `UserOptions`, `Help`, `Configuration` etc. are really meaningful names outside the navbar context.

What about registering all these as `navbar.UserOptions`, `navbar.Help`, etc.?

himdel commented on app/javascript/miq-redux/store.js:
```diff
@@ -9,6 +9,21 @@ const initialState = {};
+
+const nPrefix = '@@notifications/';
+
+const notificationReducer = (state = notificationInitialState, action) => {
```
wrong file for this, our store implementation should probably not include notification-specific reducers.

Don't we already have a way for a react component to register a reducer and initial state, instead of hardcoding it in our store?

Hyperkid123 replied:
I think that stores like this which have global scope and will be always present no matter what screen is on display, can be directly in initial state. Yes maybe we can have some file with global stores but i see no problem having them in initial state. You will add them only once and you never want to remove them.

himdel replied:
Oh, my issue is more with this happening in a surprising place, not with having the data be part of the initial state, that part is fine.

(It's about where the code lives, not when the code gets called.)

But, if we currently don't have a mechanism to do that, no issue here, no point creating one in this PR :).

fhlavac commented:
- converted ids converted to string
- added registering top navbar components with "navbar." prefix
- moved notificationReducer from store to a separate file
@himdel please review :)

himdel commented:
Thanks @fhlavac, the new changes look good :+1:
Kicking travis just to be sure with the recent changes...

skateman commented:
greeeen

skateman commented:
Okkay, let's review this...

himdel commented:
```
Uncaught Error: Attempt to create instance of RightSection. Blueprint RightSection is not defined!
```

@fhlavac pretty sure you also have to fix this one...

```
app/views/layouts/_right_section.html.haml
1:= react('RightSection', 
```

skateman commented:
Yeah
![Screenshot from 2019-12-02 16-42-46](https://user-images.githubusercontent.com/649130/69972905-ccf68200-1522-11ea-96b9-699a219f223e.png)

Hyperkid123 commented on app/javascript/packs/component-definitions-common.js:
```diff
@@ -52,26 +53,31 @@ ManageIQ.component.addReact('Breadcrumbs', Breadcrumbs);
 ManageIQ.component.addReact('Quadicon', Quadicon);
 ManageIQ.component.addReact('RemoveCatalogItemModal', RemoveCatalogItemModal);
 ManageIQ.component.addReact('ReportDataTable', ReportDataTable);
+ManageIQ.component.addReact('navbar.RightSection', navbar.RightSection);
```
Here it has to be called
```haml
= react('navbar.RightSection', { :foo => bar})
```

fhlavac replied:
Oh, I missed it...
It's fixed :)

martinpovolny commented:
Restarted travis to recover from unrelated failures.

himdel commented:
LGTM, tested vertical menu still working, group switcher, the usual buttons .. everything looks fine :+1:

skateman commented:
:tada:

Hyperkid123 commented:
@fhlavac good job

martinpovolny commented:
:heart: :heart_eyes: :heart_decoration: :+1:
